beacon_node/network/src/sync/simple_sync.rs

```diff
@@ -244,372 +209,264 @@ impl<T: BeaconChainTypes> SimpleSync<T> {
-                start_root: first.block_root,
-                start_slot: first.slot,
-                max_headers: (last.slot - first.slot + 1).as_u64(),
-                skip_slots: 0,
-            },
-            network,
+            request_id,
+            RPCResponse::BeaconBlocks(blocks.as_ssz_bytes()),
         )
     }
 
-    /// Handle a `BeaconBlockHeaders` request from the peer.
-    pub fn on_beacon_block_headers_request(
+    /// Handle a `BeaconBlocks` request from the peer.
+    pub fn on_beacon_blocks_request(
         &mut self,
         peer_id: PeerId,
         request_id: RequestId,
-        req: BeaconBlockHeadersRequest,
-        network: &mut NetworkContext,
+        req: BeaconBlocksRequest,
     ) {
         debug!(
             self.log,
-            "BlockHeadersRequest";
+            "BeaconBlocksRequest";
             "peer" => format!("{:?}", peer_id),
-            "count" => req.max_headers,
+            "count" => req.count,
+            "start_slot" => req.start_slot,
         );
 
-        let count = req.max_headers;
+        //TODO: Optimize this
+        // Currently for skipped slots, the blocks returned could be less than the requested range.
+        // In the current implementation we read from the db then filter out out-of-range blocks.
+        // Improving the db schema to prevent this would be ideal.
 
-        // Collect the block roots.
-        let mut roots: Vec<Hash256> = self
+        let mut blocks: Vec<BeaconBlock<T::EthSpec>> = self
             .chain
             .rev_iter_block_roots()
-            .take_while(|(_root, slot)| req.start_slot <= *slot)
-            .filter(|(_root, slot)| *slot < req.start_slot + count)
-            .map(|(root, _slot)| root)
-            .collect();
-
-        roots.reverse();
-        roots.dedup();
+            .filter(|(_root, slot)| {
+                req.start_slot <= slot.as_u64() && req.start_slot + req.count > slot.as_u64()
+            })
+            .take_while(|(_root, slot)| req.start_slot <= slot.as_u64())
+            .filter_map(|(root, _slot)| {
+                if let Ok(Some(block)) = self.chain.store.get::<BeaconBlock<T::EthSpec>>(&root) {
+                    Some(block)
+                } else {
+                    warn!(
+                        self.log,
+                        "Block in the chain is not in the store";
+                        "request_root" => format!("{:}", root),
+                    );
 
-        let headers: Vec<BeaconBlockHeader> = roots
-            .into_iter()
-            .step_by(req.skip_slots as usize + 1)
-            .filter_map(|root| {
-                let block = self
-                    .chain
-                    .store
-                    .get::<BeaconBlock<T::EthSpec>>(&root)
-                    .ok()?;
-                Some(block?.block_header())
+                    None
+                }
             })
+            .filter(|block| block.slot >= req.start_slot)
```

Comment:
This needed to be here for skip slots. We could request a block at slot 10, but when pulled from the db we get block at slot 5 because the last 5 slots were skipped. This is a hack filter at the moment. Ideally we would have a better mechanism to grab canonical blocks in the future